python/ray/train/v2/_internal/execution/worker_group/state.py

```diff
@@ -0,0 +1,137 @@
+import logging
+from dataclasses import dataclass
+from typing import List
+
+import ray
+from ray.actor import ActorHandle
+from ray.train.v2._internal.execution.checkpoint.sync_actor import SynchronizationActor
+from ray.train.v2._internal.execution.worker_group.worker import Worker
+from ray.util.placement_group import PlacementGroup, remove_placement_group
+
+logger = logging.getLogger(__name__)
+
+
+class _WorkerGroupShutdownMixin:
+    """Mixin class for worker group shutdown logic."""
+
+    def _shutdown_workers(self, workers: List[Worker], patience_s: float = 5):
+        # Run the worker shutdown logic on each of the workers. This should
+        # be a non-blocking call to realize forceful shutdown after patience_s.
+        _ = [w.actor.shutdown.remote() for w in workers]
+
+        logger.debug(f"Shutting down {len(workers)} workers.")
+        if patience_s <= 0:
+            for worker in workers:
+                ray.kill(worker.actor)
+        else:
+            done_refs = [w.actor.__ray_terminate__.remote() for w in workers]
+            # Wait for actors to die gracefully.
+            _, not_done = ray.wait(
+                done_refs, num_returns=len(done_refs), timeout=patience_s
+            )
+            if not_done:
+                logger.debug("Graceful termination failed. Falling back to force kill.")
+                # If all actors are not able to die gracefully, then kill them.
+                for worker in workers:
+                    ray.kill(worker.actor)
+
+    def _shutdown_sync_actor(self, sync_actor: SynchronizationActor):
+        ray.kill(sync_actor)
```

Comment:
For example, I find this sync actor lifecycle management to be a bit out of place here. Worker group should focus on its workers, not some random utility actor.

Reply:
That's a good point, I was thinking it was there to allow us to do some sort of WorkerGroup synchronization.